test/common/thread_local/thread_local_impl_test.cc

```diff
@@ -82,6 +82,32 @@ TEST_F(ThreadLocalInstanceImplTest, All) {
   tls_.shutdownThread();
 }
 
+// Validate ThreadLocal::runOnAllThreads behavior with all_thread_complete call back.
+TEST_F(ThreadLocalInstanceImplTest, RunOnAllThreads) {
+  SlotPtr tlsptr = tls_.allocateSlot();
+
+  EXPECT_CALL(thread_dispatcher_, post(_));
+  EXPECT_CALL(main_dispatcher_, post(_));
+
+  // Ensure that the thread local call back and all_thread_complete call back are called.
+  struct {
+    uint64_t thread_local_calls_{0};
+    bool all_threads_complete_ = false;
+  } thread_local_data;
+
+  tlsptr->runOnAllThreads(
+      [&thread_local_data]() -> void { ++thread_local_data.thread_local_calls_; },
+      [&thread_local_data]() -> void {
+        EXPECT_EQ(thread_local_data.thread_local_calls_, 1);
+        thread_local_data.all_threads_complete_ = true;
+      });
+
+  EXPECT_TRUE(thread_local_data.all_threads_complete_);
+
+  tls_.shutdownGlobalThreading();
+  tls_.shutdownThread();
+}
```

Comment:
At a minimum, can you leave a TODO to get the multi-thread version test running, along with more detailed information about why your previous attempt didn't work?